src/librustc/ty/layout.rs

```diff
@@ -1111,47 +1079,50 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                     // after this point – we’ll just truncate the value we load in codegen.
                 }
 
-                // Check to see if we should use a different type for the
-                // discriminant. We can safely use a type with the same size
-                // as the alignment of the first field of each variant.
+                if min_ity.size() > gap {
+                    gap += (min_ity.size() - gap).align_to(align.abi);
+                }
+
                 // We increase the size of the discriminant to avoid LLVM copying
                 // padding when it doesn't need to. This normally causes unaligned
                 // load/stores and excessive memcpy/memset operations. By using a
                 // bigger integer size, LLVM can be sure about its contents and
                 // won't be so conservative.
-
-                // Use the initial field alignment
-                let mut ity = if def.repr.c() || def.repr.int.is_some() {
+                let ity = if def.repr.inhibit_enum_layout_opt() {
```

Comment:
Not a good idea, the old condition here is not about optimizations, it's about whether the discriminant is *being forced* with `#[repr]`.